### DISCUSSION TITLE
Add BIP85 functionality

### DISCUSSION THREAD

**Discussion by aido** [aido/app-seed-tool#6] [Ideas]
```mermaid
---
title: One Seed to rule them all - Multi wallet
---
flowchart TB
    1.1 --> |Backup| 1.2
    1 --> |BIP85 Child 0| 2.1.1
    1 --> |BIP85 Child 1| 2.1.2
    1 --> |BIP85 Child 2| 2.2.1
    1 --> |BIP85 Child 3| 2.2.2
    1 --> |BIP85 Child 4| 2.3.1
    1 --> |BIP85 Child 5| 2.3.2
    1 --> |BIP85 Child 6| 2.4.1
    1 --> |BIP85 Child 7| 2.4.2
    subgraph 1[Parent]
        direction TB
        1.1[Root Seed]
        subgraph 1.2[2-of-3 Shamir's Secret Shares]
            direction BT
            1.2.1[Share 1]
            1.2.2[Share 2]
            1.2.3[Share 3]
        end
    end
    subgraph 2[Children]
        direction TB
        subgraph 2.1[Cold Wallet]
            direction LR
            2.1.1[BIP39 #1]
            2.1.2[Password #1]
            end
            subgraph 2.2[Hardware Wallet]
            direction LR
            2.2.1[BIP39 #2]
            2.2.2[Password #2]
            end
            subgraph 2.3[Lightning Wallet]
            direction LR
            2.3.1[BIP39 #3]
            2.3.2[Password #3]
            end
            subgraph 2.4[Phone Wallet]
            direction LR
            2.4.1[BIP39 #4]
            2.4.2[Password #4]
            end
    end
```

**Comment by dzid26** [comment]
For people already having a BIP39 mnemonic, I imagine they could reuse it as a BIP85 root (https://iancoleman.io/bip39/ ShowBIP85)?  

This would be beneficial in the context of inheritance simplicity - that is at the very least the heir could discover a valid wallet by treating the root seed as BIP39 and in the future if they figure out how to do BIP85 they could find children BIP39s.  

Am I not mistaken?

**Reply by aido** [reply]
Yes.

Ian Coleman's  BIP39 Mnemonic Code Converter is a great tool but it is online so I would never use my real seeds with it. That is why I am thinking of adding BIP85 functionalty to my app, so some of the utilities offered by Ian Coleman's tool can also be performed on an airgapped and secure Ledger device.

I shall be renaming the app from `app-sskr-check` to `app-seed-utils` in the not too distant future as this will better describe the apps funtionality

**Comment by aido** [comment]
@amanone has asked what the ETA is for BIP85.

The latest pull request  LedgerHQ/app-seed-tool#5 presents an "Under Construction" screen when the BIP85 option is chosen on Stax and Flex devices . This pull request is primarily a port of the app to Stax and Flex. Once this pull request has been reviewed by Ledger, merged into v1.8.0 of the app and I am happy all is working well on all devices I will then begin work on adding the BIP85 options.

**Reply by aido** [reply]
Hi @dzid26 and @amanone,

To finally follow up on this after a very long pause in development the latest pull request to Ledger's repo adds BIP85 for all Ledger e-ink signing devices; Stax and Flex including a port of the app to the newer Nano Gen5:
https://github.com/LedgerHQ/app-seed-tool/pull/14

See comment:
https://github.com/LedgerHQ/app-seed-tool/pull/14#issuecomment-4827438919

**Comment by aido** [comment]
Ledger's competitor COLDCARD have been really pushing their BIP-85 functionality lately, demonstrating just how useful it is:

https://x.com/COLDCARDwallet/status/2079551973070946736
https://x.com/COLDCARDwallet/status/2079189702759100529
https://coldcard.com/docs/bip85/

**Comment by buzzromain** [comment]
Hi @aido,

Thank you for the update and for implementing BIP85 support for the Ledger Stax, Flex and Nano Gen5.

I see that LedgerHQ/app-seed-tool#14 is still open. Do you know whether the pull request is currently being reviewed by Ledger, and whether there is an estimated timeline for it to be merged and made available to users through Ledger Live?

I understand that the final release schedule may be outside your control, but any information about the remaining steps or current blockers would be greatly appreciated.

Thanks again for your work on this feature!

**Reply by aido** [reply]
Hi @buzzromain,

Unfortunately I cannot say when the [BIP-85 PR](https://github.com/LedgerHQ/app-seed-tool/pull/14) (which includes a Nano Gen5 port) will be reviewed or if it will ever be merged by Ledger. I have had no feedback on it in over a year. I also mentioned it on Ledger's Discord but no feedback there either. My last general communication from Ledger over eight months ago informed me that people were using the app and requesting updates (e.g. a Nano Gen5 port) but still the BIP-85 PR sits there without feedback.
I think the issue may be that the PR would need a security audit, the cost of which neither Ledger or I are willing to bear.

I think the added BIP-85 functionality would be of great benefit to Ledger as it offers yet another reason to purchase one of their devices. It, and the app in general,  feeds into one of the main reasons people purchase hardware signing devices, keeping seeds safe. This is why I mentioned that COLDCARD were pushing the same functionality recently on their own devices. They see the value of BIP-85. But for now Ledger's priorities may be elsewhere.

I was even investigating additional features like saving shards generated by the app to the Ledger Recover Key, which would encourage multiple purchases of that device too (see #56). Again, zero feedback on that idea 🙄.


For now the BIP-85 PR is available to adventurous users that are OK with sideloading apps.

**Comment by buzzromain** [comment]
Hi @aido, following up on this: tdejoigny-ledger replied on #14 a few hours ago
asking if a security audit is planned, since the changes are substantial.

Before I answer him, wanted to check with you first, since it's your call too.
My instinct is that neither of us is in a position to fund an external paid
audit, and from what you wrote back in 2023 and again last year, it sounds like
you've always expected Ledger to handle that internally rather than pay for one
ourselves. Does that match where you stand on this one?

If so, I was thinking of asking him whether Ledger can carry it out internally
the way it was for v1.8.0, rather than assuming a paid audit is the only path.
Let me know if you'd rather answer him yourself, or if you see it differently.

**Reply by aido** [reply]
Hi @buzzromain,

> My instinct is that neither of us is in a position to fund an external paid
> audit, and from what you wrote back in 2023 and again last year, it sounds like
> you've always expected Ledger to handle that internally rather than pay for one
> ourselves.

Your instinct is correct. Up to the point of the current release (v1.8.1), Ledger were kind enough to take on the auditing themselves. This was agreed with Victor Forgeoux (@vforgeoux-ledger), who was responsible for reviews at the time. After the last release, it was mentioned by Victor that Ledger might not perform the reviews in future. That was where development lay until 10 months ago when Victor, due to user requests, suggested porting the app to Nano Gen5. By the time the port was complete and included in the current PR, Victor was no longer employed by Ledger.

So, the porting aspects of the latest PR were requested by a Ledger employee, but with no agreement on how the auditing would be performed.
